kernel/kernel-api/src/main/java/io/delta/kernel/internal/snapshot/SnapshotManager.java

```diff
@@ -367,9 +462,35 @@ private SnapshotImpl getSnapshotAtInit(Engine engine)
             getLogSegmentFrom(engine, lastCheckpointOpt);
 
         return logSegmentOpt
-            .map(logSegment -> createSnapshot(
-                logSegment, engine))
-            .orElseThrow(() -> new TableNotFoundException(tablePath.toString()));
+                .map(logSegment -> getCoordinatedCommitsAwareSnapshot(
+                        engine, logSegment, Optional.empty()))
+                .orElseThrow(() -> new TableNotFoundException(tablePath.toString()));
+    }
+
+    /**
+     * This can be optimized by making snapshot hint optimization to work with coordinated commits.
+     * @see <a href="https://github.com/delta-io/delta/issues/3437">issue #3437</a>.
+     */
+    private SnapshotImpl getCoordinatedCommitsAwareSnapshot(
```

Comment:
For coordinated commit table then do we list the log dir twice? Can we not reuse the log segment we've already built? If this is because there may have been changes I'm wondering if it's possible (maybe to much additional complexity) to list from the last file we've seen and reuse existing segment.

Reply:
We must list the log dir twice otherwise there might be gap because of this [issue](https://github.com/delta-io/delta/pull/3381#discussion_r1693689307).

Reply:
So the logic should be unbackfilled commit list -> deltalog list -> filter out duplicated unbackfilled commits using the results from deltalog list otherwise there might be gap or duplicated commits.